### ISSUE TITLE
String issue?

### ISSUE THREAD

**Issue by Lorenc95** [mknzz/qb-durability#2]
/server/main.lua:18: attempt to index a string value (field 'info')

Might need some other logic as it identifies the var as a string this way

```lua
            if Player.PlayerData.items[i].info.quality == nil then
                Player.PlayerData.items[i].info.quality = 100
            end
```

Won't work.

**Comment by FjamZoo** [comment]
I fixed this issue by just readjusting the function a little bit

```
RegisterServerEvent("qb-durability:server:ItemQuality")
AddEventHandler("qb-durability:server:ItemQuality", function(data, hp)
    local src = source
    local Player = QBCore.Functions.GetPlayer(src)
    if not Player then return end
    for i=0, 41 do
        local itemSlot = Player.Functions.GetItemBySlot(i)
        if itemSlot ~= nil then
            local amount = Config.DurabilityAmount["default"]
            if Config.DurabilityAmount[itemSlot.name] ~= nil then
                amount = Config.DurabilityAmount[itemSlot.name]
            end
            if Player.PlayerData.items[i].info.quality ~= nil then
                Player.PlayerData.items[i].info.quality = Player.PlayerData.items[i].info.quality - amount
                print(itemSlot.name.." removed "..amount.." durability")
                if Player.PlayerData.items[i].info.quality <= 0.0 or Player.PlayerData.items[i].info.quality <= 0 then
                    Player.Functions.RemoveItem(itemSlot.name, 1)
                    print(itemSlot.name.." removed ")
                    return
                end
            else
                Player.PlayerData.items[i].info.quality = 100
            end
            Player.Functions.SetInventory(Player.PlayerData.items)
        end
    end
end)

**Comment by MercedessBenz** [comment]
> I fixed this issue by just readjusting the function a little bit
> 
> ```
> RegisterServerEvent("qb-durability:server:ItemQuality")
> AddEventHandler("qb-durability:server:ItemQuality", function(data, hp)
>     local src = source
>     local Player = QBCore.Functions.GetPlayer(src)
>     if not Player then return end
>     for i=0, 41 do
>         local itemSlot = Player.Functions.GetItemBySlot(i)
>         if itemSlot ~= nil then
>             local amount = Config.DurabilityAmount["default"]
>             if Config.DurabilityAmount[itemSlot.name] ~= nil then
>                 amount = Config.DurabilityAmount[itemSlot.name]
>             end
>             if Player.PlayerData.items[i].info.quality ~= nil then
>                 Player.PlayerData.items[i].info.quality = Player.PlayerData.items[i].info.quality - amount
>                 print(itemSlot.name.." removed "..amount.." durability")
>                 if Player.PlayerData.items[i].info.quality <= 0.0 or Player.PlayerData.items[i].info.quality <= 0 then
>                     Player.Functions.RemoveItem(itemSlot.name, 1)
>                     print(itemSlot.name.." removed ")
>                     return
>                 end
>             else
>                 Player.PlayerData.items[i].info.quality = 100
>             end
>             Player.Functions.SetInventory(Player.PlayerData.items)
>         end
>     end
> end)
> ```

It still has the same error

**Comment by mknzz** [comment]
> /server/main.lua:18: attempt to index a string value (field 'info')
> 
> Might need some other logic as it identifies the var as a string this way
> 
> ```lua
>             if Player.PlayerData.items[i].info.quality == nil then
>                 Player.PlayerData.items[i].info.quality = 100
>             end
> ```
> 
> Won't work.

Yeah I definitely need to find a better way to do this as a whole. I think my best bet is to actually rework the inv a little and make it built-in, instead of a separate resource.

**Comment by FjamZoo** [comment]
The thing you gotta do for scripts is to set the quality upon recieving the items
that way this should get resolved.

**Comment by mknzz** [comment]
Quick fix, goto line 328 in qb-core -> server -> player.lua

Find these lines:
```
        if itemInfo['type'] == 'weapon' and info == nil then
            info = {
                serie = tostring(QBCore.Shared.RandomInt(2) .. QBCore.Shared.RandomStr(3) .. QBCore.Shared.RandomInt(1) .. QBCore.Shared.RandomStr(2) .. QBCore.Shared.RandomInt(3) .. QBCore.Shared.RandomStr(4)),
            }
        end
```

Replace with this:
```
        if itemInfo['type'] == 'weapon' and info == nil then
            info = {
                serie = tostring(QBCore.Shared.RandomInt(2) .. QBCore.Shared.RandomStr(3) .. QBCore.Shared.RandomInt(1) .. QBCore.Shared.RandomStr(2) .. QBCore.Shared.RandomInt(3) .. QBCore.Shared.RandomStr(4)),
            }
        else 
            info = {
                quality = 100
            }
        end
```
This should resolve all errors

**Comment by Lorenc95** [comment]
> Quick fix, goto line 328 in qb-core -> server -> player.lua
> 
> Find these lines:
> 
> ```
>         if itemInfo['type'] == 'weapon' and info == nil then
>             info = {
>                 serie = tostring(QBCore.Shared.RandomInt(2) .. QBCore.Shared.RandomStr(3) .. QBCore.Shared.RandomInt(1) .. QBCore.Shared.RandomStr(2) .. QBCore.Shared.RandomInt(3) .. QBCore.Shared.RandomStr(4)),
>             }
>         end
> ```
> 
> Replace with this:
> 
> ```
>         if itemInfo['type'] == 'weapon' and info == nil then
>             info = {
>                 serie = tostring(QBCore.Shared.RandomInt(2) .. QBCore.Shared.RandomStr(3) .. QBCore.Shared.RandomInt(1) .. QBCore.Shared.RandomStr(2) .. QBCore.Shared.RandomInt(3) .. QBCore.Shared.RandomStr(4)),
>             }
>         else 
>             info = {
>                 quality = 100
>             }
>         end
> ```
> 
> This should resolve all errors

That would indeed resolve all errors, this is a persistent fix even as u now set the quality flag to a 100 on every item if it doesnt have a weapon flag, should work indeed.

Only issue i can see here is if u have an item, that u dont want to degrade, but this could be done simply by setting up a config file!

**Comment by ImpulseFPS** [comment]
> Quick fix, goto line 328 in qb-core -> server -> player.lua
> 
> Find these lines:
> 
> ```
>         if itemInfo['type'] == 'weapon' and info == nil then
>             info = {
>                 serie = tostring(QBCore.Shared.RandomInt(2) .. QBCore.Shared.RandomStr(3) .. QBCore.Shared.RandomInt(1) .. QBCore.Shared.RandomStr(2) .. QBCore.Shared.RandomInt(3) .. QBCore.Shared.RandomStr(4)),
>             }
>         end
> ```
> 
> Replace with this:
> 
> ```
>         if itemInfo['type'] == 'weapon' and info == nil then
>             info = {
>                 serie = tostring(QBCore.Shared.RandomInt(2) .. QBCore.Shared.RandomStr(3) .. QBCore.Shared.RandomInt(1) .. QBCore.Shared.RandomStr(2) .. QBCore.Shared.RandomInt(3) .. QBCore.Shared.RandomStr(4)),
>             }
>         else 
>             info = {
>                 quality = 100
>             }
>         end
> ```
> 
> This should resolve all errors

On items like marked bills crypto stick and other items that have item info this change will break that at least in my experience that i have test it broke marked bills worth info

**Comment by Lorenc95** [comment]
Indeed this breaks it, as refference check idcard phone etc. All items with
a decent description.

Op do 6 jan. 2022 12:47 schreef ImpulseFPS ***@***.***>:

**Comment by FjamZoo** [comment]
You need to give information where it was nil beforehand therefor it should look like this

This way I experience no issues with markebills, crafting nor ID cards

```
        if itemInfo['type'] == 'weapon' and info == nil then
            info = {
                serie = tostring(QBCore.Shared.RandomInt(2) .. QBCore.Shared.RandomStr(3) .. QBCore.Shared.RandomInt(1) .. QBCore.Shared.RandomStr(2) .. QBCore.Shared.RandomInt(3) .. QBCore.Shared.RandomStr(4)),
            }
        elseif info == nil then
            info = {
                quality = 100
            }
        end

**Comment by AndMan21** [comment]
https://i.imgur.com/QA50Adj.png

Im still getting the error myself
